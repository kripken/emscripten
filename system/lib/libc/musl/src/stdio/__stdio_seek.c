#include "stdio_impl.h"
#ifndef __EMSCRIPTEN__
#include <unistd.h>
#endif

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
#ifdef __EMSCRIPTEN__
	off_t ret;
	if (__wasi_syscall_ret(__wasi_fd_seek(f->fd, off, whence, &ret)))
		ret = -1;
	return ret;
#else
	return __lseek(f->fd, off, whence);
#endif // __EMSCRIPTEN__
}
