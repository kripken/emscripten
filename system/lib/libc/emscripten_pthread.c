#include <pthread.h>
#include "libc.h"
#include "pthread_impl.h"
#include "stdio_impl.h"

#ifdef __EMSCRIPTEN_PTHREADS__
// In pthreads, we must initialize the runtime at the proper time, which
// is after memory is initialized and before any userland global ctors.
// We must also keep this function alive so it is always called; without
// pthreads, if pthread_self is used then this file will be included,
// and if not then it's fine to not have this.
EM_JS(void, initPthreadsJS, (void), {
  PThread.initRuntime();
})

static void init_file_lock(FILE *f)
{
  if (f && f->lock<0) f->lock = 0;
}

// std{in,out,err}.c will override this if linked
static FILE *volatile dummy_file = 0;
weak_alias(dummy_file, __stdin_used);
weak_alias(dummy_file, __stdout_used);
weak_alias(dummy_file, __stderr_used);

// This must run before any userland ctors
// Note that ASan constructor priority is 50, and we must be higher.
EMSCRIPTEN_KEEPALIVE
__attribute__((constructor(48)))
void __emscripten_pthread_data_constructor(void) {
  initPthreadsJS();
  pthread_self()->locale = &libc.global_locale;
  init_file_lock(__stdin_used);
  init_file_lock(__stdout_used);
  init_file_lock(__stderr_used);
  libc.threaded = 1;
}
#else // !defined(__EMSCRIPTEN_PTHREADS__)
static struct pthread __main_pthread;
uintptr_t __get_tp(void) {
  return (uintptr_t)&__main_pthread;
}

__attribute__((constructor))
void __emscripten_pthread_data_constructor(void) {
  __pthread_self()->locale = &libc.global_locale;
}
#endif // !defined(__EMSCRIPTEN_PTHREADS__)
