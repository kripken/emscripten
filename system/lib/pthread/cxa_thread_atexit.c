/*
 * Copyright 2019 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 * Inspired by libcxxabi/src/cxa_thread_atexit.cpp.
 * The main reasons we don't use that version direclty are:
 * 1. We want to be able to use __cxa_thread_atexit in pure C programs
 *    where libcxxabi is not linked in at all.
 * 2. The libcxxabi relies on TLS variables, which we can't use here.
 *
 * WARNING: We specifically can't use TLS variables or pthread_once in this file
 * because they both end up calling __cxa_thread_atexit.
 *
 * TLS variables: TLS data gets freed using __cxa_thread_atexit.  This means
 * that if we use TLS to store our DtorList it will be free'd while porcessing
 * itself.  See lib/pthread/emscripten_tls_init.c.
 *
 * pthread_once: calls pthread_cleanup_push / pthread_cleanup_pop, which call
 * __cxa_thread_atexit on first use.  See src/thread/pthread_once.c.
 */
#include <assert.h>
#include <stdbool.h>
#include <stdatomic.h>
#include <pthread.h>
#include <stdlib.h>

#include "libc.h"

typedef void (*Dtor)(void*);

typedef struct DtorList {
  Dtor dtor;
  void* obj;
  struct DtorList* next;
} DtorList;

void run_dtors(void* arg) {
  DtorList* dtors = (DtorList*)arg;
  DtorList* head;
  while ((head = dtors)) {
    dtors = head->next;
    head->dtor(head->obj);
    free(head);
  }
}

static pthread_key_t key;
static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

static void ensure_key() {
  // See above for why we can't use pthread_once.
  pthread_mutex_lock(&mutex);
  static atomic_bool key_created = false;
  if (!key_created) {
    pthread_key_create(&key, run_dtors);
    key_created = true;
  }
  pthread_mutex_unlock(&mutex);
}

int __cxa_thread_atexit_impl(Dtor dtor, void* obj, void* dso_symbol) {
  ensure_key();
  DtorList* old_head = pthread_getspecific(key);
  DtorList* head = (DtorList*)(malloc(sizeof(DtorList)));
  assert(head);
  head->dtor = dtor;
  head->obj = obj;
  head->next = old_head;
  pthread_setspecific(key, head);
  return 0;
}

weak_alias(__cxa_thread_atexit_impl, __cxa_thread_atexit);
