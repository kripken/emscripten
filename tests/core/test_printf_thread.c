#include <assert.h>
#include <pthread.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../system/lib/libc/musl/src/include/features.h"
#include "../../system/lib/libc/musl/src/internal/stdio_impl.h"
#include "../../system/lib/libc/musl/src/internal/libc.h"

pthread_t thread[2];

char *char_repeat(int n, char c) {
  char *dest = malloc(n + 1);
  memset(dest, c, n);
  dest[n] = '\0';
  return dest;
}

void *thread_main(void *arg) {
  char *msg = char_repeat(100, 'a');
  for (int i = 0; i < 10; ++i)
    printf("%s\n", msg);
  free(msg);
  return 0;
}

int main() {
  printf("stdin lock: %d\n", stdin->lock);
  printf("stdout lock: %d\n", stdout->lock);
  printf("stderr lock: %d\n", stderr->lock);
  printf("threaded libc: %d\n", libc.threaded);
  int rc = pthread_create(&thread[0], NULL, thread_main, NULL);
  assert(rc == 0);
  rc = pthread_create(&thread[1], NULL, thread_main, NULL);
  assert(rc == 0);
  void *thread_rtn = 0;
  rc = pthread_join(thread[0], &thread_rtn);
  assert(rc == 0);
  assert(thread_rtn == 0);
  rc = pthread_join(thread[1], &thread_rtn);
  assert(rc == 0);
  assert(thread_rtn == 0);

  return 0;
}
