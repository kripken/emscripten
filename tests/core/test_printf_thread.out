stdin lock: 0
stdout lock: 0
stderr lock: 0
threaded libc: 1
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
